modules/data_models.py

```diff
@@ -0,0 +1,363 @@
+            blue_plus = f"\33[94m{'+' * entry[4]}"
+            report_line = (
+                f"{day} {blue_plus}{red_plus} "
+                f"\33[0m{entry[4]}{TEMPERATURE_UNIT}-{entry[3]}{TEMPERATURE_UNIT}"
+            )
+            print(report_line)
+
+
+class YearData:
+    """
+    This class holds data of an entire year
+    """
+
+    def __init__(self, year, path=WEATHER_FILES_DIR):
+        """
+        Initializes the members year, path and calls populate to fill months_data list
+        Args:
+            year(str or int):   Value containing 4 digit year e.g '2004'
+            path(str):  Value containing path to weather files e.g 'weatherfiles/'
+        """
+        self.months_data = []
+        self.year = year
+        self.path = path
+        self.populate()
+
+    def populate(self):
+        """
+        Initializes the months_data member with MonthData objects
+        Returns:
+            None
+        """
+        for month in range(1, 13):
+            month_data = MonthData(self.year, get_month_name(month)[0:3], self.path)
+            self.months_data.append(month_data)
+
+    def get_max_extremes_with_date(self):
+        """
+        Returns maximum Highest temperature, minimum lowest temperature and
+        maximum max humidity for the whole year with relevant dates
+        Args:
+        Returns:
+            (list or None): list containing max extremes with dates
+                            [
+                                [max_temperature(int), date(str)],
+                                [min_temperature(int), date(str)],
+                                [max_humidity(int), date(str)]
+                            ]
+                            Or
+                            None if there is no data for the year
+        """
+
+        if not self.months_data:
+            return None
+        max_temperature = {"value": -1000, "date": ""}
+        min_temperature = {"value": 1000, "date": ""}
+        max_humidity = {"value": -1000, "date": ""}
+        for month_data in self.months_data:
+            month_max_extremes = month_data.get_max_extremes_with_date()
+
+            max_temperature = max(
+                max_temperature,
+                month_max_extremes[0],
+                key=lambda x: x["value"] or -1000,
+            )
+
+            min_temperature = min(
+                min_temperature, month_max_extremes[1], key=lambda x: x["value"] or 1000
+            )
+
+            max_humidity = max(
+                max_humidity, month_max_extremes[2], key=lambda x: x["value"] or -1000
+            )
+        max_temperature["value"] = f"{max_temperature['value']}{TEMPERATURE_UNIT}"
+        min_temperature["value"] = f"{min_temperature['value']}{TEMPERATURE_UNIT}"
+        max_humidity["value"] = f"{max_humidity['value']}{HUMIDITY_UNIT}"
+        return [max_temperature, min_temperature, max_humidity]
+
+    def generate_extremes_report(self):
```

Comment:
Similar to above comments generate reports should be the responsibility of the report generator, not the MonthData Class. The responsibility of MonthData class should be to manipulate the data, dave it and return proper values to the callers.

Reply:
Done